pkg/config/config.go

```diff
@@ -255,60 +243,72 @@ func ConvertToInfraParams(p any) InfraParams {
 }
 
 func (a *Application) MergeDefaults(other Defaults) {
-	a.Defaults.ExecutionUnit.Merge(other.ExecutionUnit)
-	a.Defaults.Expose.Merge(other.Expose)
-	a.Defaults.PersistFs.Merge(other.PersistFs)
-	a.Defaults.PersistKv.Merge(other.PersistKv)
-	a.Defaults.PersistOrm.Merge(other.PersistOrm)
-	a.Defaults.PersistRedisCluster.Merge(other.PersistRedisCluster)
-	a.Defaults.PersistRedisNode.Merge(other.PersistRedisNode)
-	a.Defaults.PersistSecrets.Merge(other.PersistSecrets)
-	a.Defaults.PubSub.Merge(other.PubSub)
-	a.Defaults.StaticUnit.Merge(other.StaticUnit)
-	a.Defaults.Config.Merge(other.Config)
+	a.Defaults.ExecutionUnit.ApplyDefaults(other.ExecutionUnit)
+	a.Defaults.Expose.ApplyDefaults(other.Expose)
+	a.Defaults.PersistFs.ApplyDefaults(other.PersistFs)
+	a.Defaults.PersistKv.ApplyDefaults(other.PersistKv)
+	a.Defaults.PersistOrm.ApplyDefaults(other.PersistOrm)
+	a.Defaults.PersistRedisCluster.ApplyDefaults(other.PersistRedisCluster)
+	a.Defaults.PersistRedisNode.ApplyDefaults(other.PersistRedisNode)
+	a.Defaults.PersistSecrets.ApplyDefaults(other.PersistSecrets)
+	a.Defaults.PubSub.ApplyDefaults(other.PubSub)
+	a.Defaults.StaticUnit.ApplyDefaults(other.StaticUnit)
+	a.Defaults.Config.ApplyDefaults(other.Config)
 }
 
-func (cfg *KindDefaults) Merge(other KindDefaults) {
-	if other.Type != "" && cfg.Type == "" {
-		cfg.Type = other.Type
+func (cfg *KindDefaults) ApplyDefaults(dflt KindDefaults) {
+	if cfg.Type == "" {
+		cfg.Type = dflt.Type
 	}
 	if cfg.InfraParamsByType == nil {
 		cfg.InfraParamsByType = make(map[string]InfraParams)
 	}
-	for name, unit := range other.InfraParamsByType {
-		paramsByType := cfg.InfraParamsByType[name]
-		cfg.InfraParamsByType[name] = paramsByType.Merge(unit)
+	for name, unit := range dflt.InfraParamsByType {
+		params := cfg.InfraParamsByType[name]
+		params.ApplyDefaults(unit)
+		cfg.InfraParamsByType[name] = params
 	}
 }
 
 var (
 	MaxDepth = 32
 )
 
-// Merge recursively merges the src and dst maps. Key conflicts are resolved by
-// preferring src, or recursively descending, if both src and dst are maps.
-func (src InfraParams) Merge(dst InfraParams) map[string]interface{} {
-	return merge(dst, src, 0)
+// ApplyDefaults applies the defaults to params
+func (params *InfraParams) ApplyDefaults(dflt InfraParams) {
+	if len(dflt) == 0 {
+		return
+	}
+	if *params == nil {
+		*params = make(InfraParams)
+	}
+	merge(*params, dflt, 0)
 }
 
-func merge(dst, src map[string]interface{}, depth int) map[string]interface{} {
+// merge applies all k,v pairs from src into dst. If the value is a map, it will
+// try to recusively merge those as well. When keys conflict, dst will win out
+// because this is used for ApplyDefaults where dst is the specific annotation's
+// values and src is the default values.
+func merge(dst, src map[string]interface{}, depth int) {
 	if depth > MaxDepth {
 		panic("too deep!")
 	}
```

Comment:
nit: I realize you didnt write this, but should we change the panic message. Kind of a funny message and shouldnt ever happen, but would be weird for a user to see that